tests/ui/pin-ergonomics/borrow-mut-xor-share.rs

```diff
@@ -0,0 +1,62 @@
+#![feature(pin_ergonomics)]
+#![allow(dead_code, incomplete_features)]
+
+// For now, we move the place in `&pin mut place` to ensure soundness.
+// In the next step, we borrow the place instead of moving it, after that we
+// have to makes sure `&pin mut place` and `&pin const place` cannot violate
+// the mut-xor-share rules.
+
+use std::pin::Pin;
+
+struct Foo;
+
+fn foo_mut(_: &mut Foo) {
+}
+
+fn foo_ref(_: &Foo) {
+}
+
+fn foo_pin_mut(_: Pin<&mut Foo>) {
+}
+
+fn foo_pin_ref(_: Pin<&Foo>) {
+}
+
+fn bar() {
+    let foo = Foo;
+    foo_pin_mut(&pin mut foo); // ok because `foo` is moved`
+
+    let mut foo = Foo;
+    let x = &pin mut foo;
+    foo_pin_ref(&pin const foo); //~ ERROR borrow of moved value: `foo`
+    foo_pin_mut(&pin mut foo); // ok
+    foo_ref(&foo); //~ ERROR borrow of moved value: `foo`
+    foo_mut(&mut foo); // ok
```

Comment:
Maybe I'm missing something, but if we're doing a move to ensure soundness, as the comments indicate, then shouldn't the lines marked `// ok` be failing here, just as they would with `pin!`?  Certainly, if `Foo` were `?Unpin`, we could not allow `&mut foo` here.  Why does this work here?

Reply:
I think that's because only the first borrow-of-moved-value error is emitted regarding the same moved value.

[Playgraound](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=e731646ad1476d31a967099d02bacb12).

Reply:
I rewrote these tests to avoid this misleading absent error message.

Reply:
Also, only moving the `!Unpin` typed values to support more flexible usages.